### PR TITLE
feat: genefuse wrapper

### PR DESCRIPTION
**Pull request by Smeds**
### Description

Wrapper used to run genefuse

### QC

For all wrappers added by this PR, I made sure that

* [x] there is a test case which covers any introduced changes,
* [x] `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* [x] either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* [x] rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* [x] all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* [x] wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* [x] all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* [x] `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* [x] temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* [x] the `meta.yaml` contains a link to the documentation of the respective tool or command,
* [x] `Snakefile`s pass the linting (`snakemake --lint`),
* [x] `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* [x] Python wrapper scripts are formatted with [black](https://black.readthedocs.io).

### CODE REVIEW

File: bio/genefuse/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  - Patrik Smeds
+input:
+  - fastq files
+  - gene fuse settings files
+  - refeference genome
+output:
+  - txt file with fusions
+  - html report
+notes: |
+  * For more inforamtion see, https://github.com/OpenGene/GeneFuse
```
Comment (fgvieira) [inline bio/genefuse/meta.yaml:13]: url: https://github.com/OpenGene/GeneFuse
authors:
  - Patrik Smeds
input:
  - fastq files
  - gene fuse settings files
  - refeference genome
output:
  - txt file with fusions
  - html report
notes: |
  * The `extra` param allows for additional program arguments.

File: bio/genefuse/environment.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+channels:
+  - bioconda
+  - conda-forge
+  - defaults
+dependencies:
+  - genefuse=0.6.1
```
Comment (fgvieira) [inline bio/genefuse/environment.yaml:6]: - genefuse=0.6

File: bio/genefuse/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    "(genefuse "
+    "-r {snakemake.input.reference} "
+    "-t {snakemake.threads} "
+    "-f {snakemake.input.config} "
+    "-1 {snakemake.input.fastq1} "
+    "-2 {snakemake.input.fastq2} "
+    "-h {snakemake.output.html} "
+    "{extra} > "
+    "{snakemake.output.fusions}) "
+    "{log}"
```
Comment (fgvieira) [inline bio/genefuse/wrapper.py:25]: "{snakemake.output.fusions} "
    "{log})"

**Comment by Smeds** [comment]
@fgvieira 

neither json or html are required! But from my understanding both will always be generated, if you don't use flags -h and -j you will end up with default names. Since the files always will be generated I thought it would be good pratices to force the user to specify the names, or set them to temp if snakemake should remove them.

**Comment by fgvieira** [comment]
So the `html` and `json` files are generated even if options `-h`/`-j` are not specified?
What if the wrapper always sets `-h` and `-j` to some temporary files and, if the user specified them as outputs, copy them to  their final destination?

**Comment by Smeds** [comment]
I

> So the `html` and `json` files are generated even if options `-h`/`-j` are not specified?
> What if the wrapper always sets `-h` and `-j` to some temporary files and, if the user specified them as outputs, copy them to their final destination?

They will always be generated from what I can see. I can change to temporary files and then maybe add a check that at least one of html or json is specified?

**Comment by fgvieira** [comment]
> ...and then maybe add a check that at least one of html or json is specified?

But why would we need to enforce that, if all works fine if `-h` and `-j` are not specififed?
If the user specifies them as rule outputs, then the wrapper copies them to their final destination. If none of them are specified, then they are automatically deleted when the rule ends (since they are temp).

**Comment by Smeds** [comment]
> > ...and then maybe add a check that at least one of html or json is specified?
> 
> But why would we need to enforce that, if all works fine if `-h` and `-j` are not specififed? If the user specifies them as rule outputs, then the wrapper copies them to their final destination. If none of them are specified, then they are automatically deleted when the rule ends (since they are temp).

Couldn't that give rise to confusing errors? Let say someone use the wrapper and specify output without using a keyword, in that case the program will run properly and snakemake would generate a missing output file error.

**Comment by fgvieira** [comment]
I think that would be ok, since the user would have to figure out what is wrong.
But maybe we could also hear @johanneskoester... :smile:

**Comment by Smeds** [comment]
@fgvieira new pull-request where output is written to a temp folder and then copied to the specifed output path

File: bio/genefuse/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+input:
+  - fastq files
+  - gene fuse settings files
+  - refeference genome
+output:
+  - txt file with fusions
+  - html report
+  - json report
+notes: |
+  * For more inforamtion see, https://github.com/OpenGene/GeneFuse
```
Comment (fgvieira) [inline bio/genefuse/meta.yaml:14]:
```suggestion
  * The `extra` param allows for additional program arguments.
```

File: bio/genefuse/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    txt_path = f"{tempdir}/gene_fuse_fusions.txt"
+    shell(
+        "(genefuse "
+        "-r {snakemake.input.reference} "
+        "-t {snakemake.threads} "
+        "-f {snakemake.input.config} "
+        "-1 {snakemake.input.fastq1} "
+        "-2 {snakemake.input.fastq2} "
+        f"-h {html_path} "
+        f"-j {json_path} "
```
Comment (fgvieira) [inline bio/genefuse/wrapper.py:28]:
```suggestion
        "-h {html_path} "
        "-j {json_path} "
```

File: bio/genefuse/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        "{extra} > "
+        f"{txt_path}) "
```
Comment (fgvieira) [inline bio/genefuse/wrapper.py:30]:
```suggestion
        "{txt_path}) "
```

File: bio/genefuse/wrapper.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        "-2 {snakemake.input.fastq2} "
+        "-h {html_path} "
+        "-j {json_path} "
+        "{extra} > "
+        f"{txt_path}) "
+        "{log}"
+    )
+
+    if snakemake.output.get("html", None):
+        shell(f"mv {html_path} {snakemake.output.html}")
```
Comment (fgvieira) [inline bio/genefuse/wrapper.py:35]:
```suggestion
        shell("mv {html_path} {snakemake.output.html}")
```

File: bio/genefuse/wrapper.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    if snakemake.output.get("json", None):
+        shell(f"mv {json_path} {snakemake.output.json}")
```
Comment (fgvieira) [inline bio/genefuse/wrapper.py:38]:
```suggestion
        shell("mv {json_path} {snakemake.output.json}")
```

File: bio/genefuse/wrapper.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    if snakemake.output.get("fusions", None):
+        shell(f"mv {txt_path} {snakemake.output.fusions}")
```
Comment (fgvieira) [inline bio/genefuse/wrapper.py:41]:
```suggestion
        shell("mv {txt_path} {snakemake.output.fusions}")
```

File: bio/genefuse/environment.yaml (hunk 2 above)
Comment (fgvieira) [inline bio/genefuse/environment.yaml:6]:
```suggestion
  - genefuse=0.6
```

**Review by fgvieira — APPROVED** [review]

**Review by johanneskoester — APPROVED** [review]
Excellent work @Smeds and @fgvieira!

**Pull request merged**